File: Java/extension/WaveletMatrix.java
package extension;

import java.util.Arrays;
import java.util.stream.IntStream;

final class SuccinctIndexableDictionary {
	private final int blk;
	private final int[] bit, sum;
	SuccinctIndexableDictionary(final int len) {
		blk = (len + 31) >> 5;
		bit = new int[blk];
		sum = new int[blk];
	}
	final void set(final int k){ bit[k >> 5] |= 1 << (k & 31); }
	final void build() {
		sum[0] = 0;
		for(int i = 0; ++i < blk;) {
			sum[i] = sum[i - 1] + Integer.bitCount(bit[i - 1]);
		}
	}
	final boolean get(final int k){ return ((bit[k >> 5] >> (k & 31)) & 1) == 1; }
	final int rank(final int k){ return (sum[k >> 5] + Integer.bitCount(bit[k >> 5] & ((1 << (k & 31)) - 1))); }
	final int rank(final boolean val, final int k){ return val ? rank(k) : k - rank(k); }
}
final class WaveletMatrixBeta {
	private final int log;
	private final SuccinctIndexableDictionary[] matrix;
	private final int[] mid;
	WaveletMatrixBeta(long[] arr, final int log) {
		final int len = arr.length;
		this.log = log;
		matrix = new SuccinctIndexableDictionary[log];
		mid = new int[log];
		final long[] l = new long[len], r = new long[len];
		for(int level = log; --level >= 0;) {
			matrix[level] = new SuccinctIndexableDictionary(len + 1);
			int left = 0, right = 0;
			for(int i = 0; i < len; ++i) {
				if(((arr[i] >> level) & 1) == 1) {
					matrix[level].set(i);
					r[right++] = arr[i];
				} else {
					l[left++] = arr[i];
				}
			}
			mid[level] = left;
			matrix[level].build();
			final long[] tmp = new long[len];
			System.arraycopy(arr, 0, tmp, 0, len);
			System.arraycopy(l, 0, arr, 0, len);
			System.arraycopy(tmp, 0, l, 0, len);
			for(int i = 0; i < right; ++i) {
				arr[left + i] = r[i];
			}
		}
	}
	final NumPair succ(final boolean f, final int l, final int r, final int level){ return new NumPair(matrix[level].rank(f, l) + mid[level] * (f ? 1 : 0), matrix[level].rank(f, r) + mid[level] * (f ? 1 : 0)); }
	final long access(int k) {
		long ret = 0;
		for(int level = log; --level >= 0;) {
			final boolean f = matrix[level].get(k);
			if(f) {
				ret |= 1L << level;
			}
			k = matrix[level].rank(f, k) + mid[level] * (f ? 1 : 0);
		}	
		return ret;
	}
	final int rank(final long x, int r) {
		int l = 0;
		for(int level = log; --level >= 0;) {
			final var p = succ(((x >> level) & 1) == 1, l, r, level);
			l = p.first.intValue();
			r = p.second.intValue();
		}
		return r - l;
	}
	final long kthMin(int l, int r, int k) {
		if(Utility.scope(0, k, r - l - 1)) {
			throw new IndexOutOfBoundsException();
		}
		long ret = 0;
		for(int level = log; --level >= 0;) {
			final int cnt = matrix[level].rank(false, r) - matrix[level].rank(false, l);
			final boolean f = cnt <= k;
			if(f) {
				ret |= 1 << level;
				k -= cnt;
			}
			final var p = succ(f, l, r, level);
			l = p.first.intValue();
			r = p.second.intValue();
		}
		return ret;
	}
	final long kthMax(final int l, final int r, final int k){ return kthMin(l, r, r - l - k - 1); }
	final int rangeFreq(int l, int r, final long upper) {
		int ret = 0;
		for(int level = log; --level >= 0;) {
			final boolean f = ((upper >> level) & 1) == 1;
			if(f) {
				ret += matrix[level].rank(false, r) - matrix[level].rank(false, l);
			}
			final var p = succ(f, l, r, level); 
			l = p.first.intValue();
			r = p.second.intValue();
		}
		return ret;
	}
	final int rangeFreq(final int l, final int r, final long lower, final long upper){ return rangeFreq(l, r, upper) - rangeFreq(l, r, lower); }
	final long prevValue(final int l, final int r, final long upper) {
		final int cnt = rangeFreq(l, r, upper);
		return cnt == 0 ? -1 : kthMin(l, r, cnt - 1);
	}
	final long nextValue(final int l, final int r, final long lower) {
		final int cnt = rangeFreq(l, r, lower);
		return cnt == r - l ? -1 : kthMin(l, r, cnt);
	}
}
final class WaveletMatrix {
	private final WaveletMatrixBeta mat;
	private final long[] ys;
	WaveletMatrix(final long[] arr, final int log) {
		ys = Arrays.stream(arr).sorted().distinct().toArray();
		final long[] t = new long[arr.length];
		IntStream.range(0, arr.length).forEach(i -> t[i] = get(arr[i]));
		mat = new WaveletMatrixBeta(t, log);
	}
	private int lowerBound(final long[] arr, final long x) {
		final int id = Arrays.binarySearch(arr, x);
		return id < 0 ? -(id - 1) : id;
	}
	private final int get(final long x){ return lowerBound(ys, x); }
	final long access(final int k){ return ys[(int) mat.access(k)]; }
	final int rank(final long x, final int r) {
		final var pos = get(x);
		if(pos == ys.length || ys[pos] != x) {
			return 0;
		}
		return mat.rank(pos, r);
	}
	final long kthMin(final int l, final int r, final int k){ return ys[(int) mat.kthMin(l, r, k)]; }
	final long kthMax(final int l, final int r, final int k){ return ys[(int) mat.kthMax(l, r, k)]; }
	final int rangeFreq(final int l, final int r, final long upper){ return mat.rangeFreq(l, r, get(upper)); }
	final int rangeFreq(final int l, final int r, final long lower, final long upper){ return mat.rangeFreq(l, r, get(lower), get(upper)); }
	final long prevValue(final int l, final int r, final long upper) {
		final var ret = mat.prevValue(l, r, get(upper));
		return ret == -1 ? -1 : ys[(int) ret];
	}
	final long nextValue(final int l, final int r, final long lower) {
		final var ret = mat.nextValue(l, r, get(lower));
		return ret == -1 ? -1 : ys[(int) ret];
	}
}

File: Nim/WaveletMatrix.nim
import bitops, algorithm
proc `|`(x: int, y: int): int = x or y
proc `&`(x: int, y: int): int = x and y
proc `>>`(x: int, y: int): int = x shr y
proc `<<`(x: int, y: int): int = x shl y
proc `|=`(x: var int, y: int): void = x = x | y
proc `?`(x: bool): int =
    if x: 1
    else: 0
type Pair[F, S] = ref object
    first*: F
    second*: S
proc initPair[F, S](f: F, s: S): Pair[F, S] {.inline} = Pair[F, S](first: f, second: s)
type SIDict = ref object
    blk: int
    bit, sum: seq[int]
proc initSIDict(n: int): SIDict {.inline} =
    var sid = new SIDict
    sid.blk = (n + 31) >> 5
    sid.bit.setLen((n + 31) >> 5)
    sid.sum.setLen((n + 31) >> 5)
    sid
proc set(sid: SIDict, i: int) {.inline} = sid.bit[i >> 5] |= 1 << (i & 31)
proc build(sid: SIDict) {.inline} =
    sid.sum[0] = 0
    for i in 0..<sid.blk - 1: sid.sum[i + 1] = sid.sum[i] + popcount(sid.bit[i])
proc `[]`(sid: SIDict, i: int): bool {.inline} = ((sid.bit[i >> 5] >> (i & 31)) & 1) == 1
proc rank(sid: SIDict, i: int): int {.inline} = (sid.sum[i >> 5] + popcount(sid.bit[i >> 5] & ((1 << (i & 31)) - 1)))
proc rank(sid: SIDict, val: bool, i: int): int {.inline} =
    if val: sid.rank(i)
    else: i - sid.rank(i)
type WMBeta = ref object
    log: int
    mat: seq[SIDict]
    mid: seq[int]
proc initWMBeta(arr: seq[int], log: int): WMBeta {.inline} =
    var wmb = new WMBeta
    var a = arr
    let len = len(a)
    wmb.log = log
    wmb.mat.setLen(log)
    wmb.mid.setLen(log)
    var l, r: seq[int]
    l.setLen(len)
    r.setLen(len)
    for level in countdown(log - 1, 0):
        wmb.mat[level] = initSIDict(len + 1)
        var left, right: int
        left = 0
        right = 0
        for i in 0..<len:
            if ((a[i] >> level) & 1) == 1:
                wmb.mat[level].set(i)
                r[right] = a[i]
                right += 1
            else:
                l[left] = a[i]
                left += 1
        wmb.mid[level] = left;
        wmb.mat[level].build();
        swap(a, l)
        for i in 0..<right: a[left + i] = r[i]
    wmb
proc `[]`(wmb: WMBeta, i: int): int {.inline} =
    var res, k: int
    res = 0
    k = i
    for level in countdown(wmb.log - 1, 0):
        let f = wmb.mat[level][k]
        if f: res |= 1 << level
        k = wmb.mat[level].rank(f, k) + wmb.mid[level] * ?f
    res
proc succ(wmb: WMBeta, f: bool, l: int, r: int, level: int): Pair[int, int] {.inline} = initPair(wmb.mat[level].rank(f, l) + wmb.mid[level] * ?f, wmb.mat[level].rank(f, r) + wmb.mid[level] * ?f)
proc rank(wmb: WMBeta, x: int, id: int): int {.inline} =
    var l = 0
    var r = id
    for level in countdown(wmb.log - 1, 0):
        let p = wmb.succ((x >> level) & 1 == 1, l, r, level)
        l = p.first
        r = p.second
    r - l
proc kthMin(wmb: WMBeta, a: int, b: int, i: int): int {.inline} =
    assert 0 <= i and i < b - a
    var l, r, k, ret: int
    l = a
    r = b
    k = i
    ret = 0
    for level in countdown(wmb.log - 1, 0):
        let cnt = wmb.mat[level].rank(false, r) - wmb.mat[level].rank(false, l)
        let f = cnt <= k
        if f:
            ret |= 1 << level
            k -= cnt
        let p = wmb.succ(f, l, r, level)
        l = p.first
        r = p.second
    ret
proc kthMax(wmb: WMBeta, l: int, r: int, k: int): int {.inline} = wmb.kthMin(l, r, r - l - k - 1)
proc rangeFreq(wmb: WMBeta, a: int, b: int, upper: int): int {.inline} =
    var l, r, ret: int
    l = a
    r = b
    ret = 0
    for level in countdown(wmb.log - 1, 0):
        let f = upper >> level & 1 == 1
        if f: ret += wmb.mat[level].rank(false, r) - wmb.mat[level].rank(false, l)
        let p = wmb.succ(f, l, r, level)
        l = p.first
        r = p.second
    ret
proc rangeFreq(wmb: WMBeta, l: int, r: int, lower: int, upper: int): int {.inline} = wmb.rangeFreq(l, r, upper) - wmb.rangeFreq(l, r, lower)
proc prev(wmb: WMBeta, l: int, r: int, upper: int): int {.inline} =
    let cnt = wmb.rangeFreq(l, r, upper)
    if cnt == 0: -1
    else: wmb.kthMin(l, r, cnt - 1)
proc next(wmb: WMBeta, l: int, r: int, lower: int): int {.inline} =
    let cnt = wmb.rangeFreq(l, r, lower)
    if cnt == 0: r - l
    else: wmb.kthMin(l, r, cnt)
type WaveletMatrix = ref object
    mat: WMBeta
    ys: seq[int]
proc get(wm: WaveletMatrix, x: int): int {.inline} = lowerBound(wm.ys, x)
proc uniq(a: openArray[int]): seq[int] {.inline} =
    var res: seq[int]
    var j = -1
    if len(a) > 0:
        j = 0
        res.add(a[0])
    for i, el in a:
        if a[j] == el: continue
        j = i
        res.add(el)
    res
proc initWaveletMatrix*(a: seq[int], log: int = 20): WaveletMatrix {.inline} =
    var wm = new WaveletMatrix
    wm.ys = sorted(a).uniq
    var t = newSeq[int](len(a))
    for i, el in a: t[i] = wm.get(el)
    wm.mat = initWMBeta(t, log)
    wm
proc `[]`*(wm: WaveletMatrix, i: int): int {.inline} = wm.ys[wm.mat[i]]
proc rank*(wm: WaveletMatrix, r: int, x: int): int {.inline} =
    let pos = wm.get(x)
    if pos == len(wm.ys) or wm.ys[pos] != x: 0
    else: wm.mat.rank(pos, r)
proc rank*(wm: WaveletMatrix, l: int, r: int, x: int): int {.inline} = wm.rank(r, x) - wm.rank(l, x)
proc kthMin*(wm: WaveletMatrix, l: int, r: int, k: int): int {.inline} = wm.ys[wm.mat.kthMin(l, r, k)]
proc kthMax*(wm: WaveletMatrix, l: int, r: int, k: int): int {.inline} = wm.ys[wm.mat.kthMax(l, r, k)]
proc rangeFreq*(wm: WaveletMatrix, l: int, r: int, upper: int): int {.inline} = wm.mat.rangeFreq(l, r, wm.get(upper))
proc rangeFreq*(wm: WaveletMatrix, l: int, r: int, lower: int, upper: int): int {.inline} = wm.mat.rangeFreq(l, r, wm.get(lower), wm.get(upper))
proc prev*(wm: WaveletMatrix, l: int, r: int, upper: int): int {.inline} =
    let ret = wm.mat.prev(l, r, wm.get(upper))
    if ret == -1: -1
    else: wm.ys[ret]
proc next*(wm: WaveletMatrix, l: int, r: int, lower: int): int {.inline} =
    let ret = wm.mat.next(l, r, wm.get(lower))
    if ret == -1: -1
    else: wm.ys[ret]